## Gradient Descent

* [Importing dependencies](#idp)
* [Definining the Gradient Descent Function](#gd)
* [Executing the function with different parameters](#dp)

Gradient Descent aims to find the optimal values of parameters (weights and bias) by minimzing a loss function. 
1) Initialize random parameter values

2) Calculate the error between the predictions and the actual target values

3) Obtain the derivative of the loss function with respect to the parameter

4) Update the parameter values based on the derivative. Step size = derivative x learning rate

5) Repeat steps 2-4 until step size reaches close to 0 or number of steps hits the defined maximum

### Importing Dependencies <a id = "idp">

In [1]:
import numpy as np

In [ ]:
# Declaring test data
X = np.array([1, 2, 3, 4, 5])
Y = np.array([5, 7, 9, 11, 13]) # Target: y = 2x + 3

### Defining the Gradient Descent Function <a id = "gd">

In [47]:
def gradient_descent(x, y, iterations, learning_rate):
    # Start with 1 weight and 1 bias
    w = 0.0
    b = 0.0
    n = len(x)
    
    for i in range(iterations):
        # Forward Pass to get initial prediction
        y_predicted = w * x + b
        
        # Get the difference between the predicted and target value
        # MSE is used in this circumstance
        cost = (1/n) * sum([val**2 for val in (y - y_predicted)])

        # Calculate the derivatives
        dw = -(2/n) * sum(x * (y - y_predicted))
        db = -(2/n) * sum(y - y_predicted)
        #Derivative of w: 
        #d(Cost)/d(w) = d(1/n * sum(y-y pred)^2) with respect to w
        #             = 1/n*(sum(2(y-y pred)) x d(y-y pred)
        #             = 1/n*(sum(2(y-y pred)) x d(y - (wx + b))
        #             = 1/n*(sum(2(y-y pred)) x d(y - wx - b)
        #             = 1/n*(sum(2(y-y pred)) x (0 - x * 0 - 0)
        #             = 1/n*(sum(2(y-y pred)) x (-x)
        #             = -2/n*(sum(x(y-y pred)))
        
        # Update parameters based on (derivative x learning rate)
        w = w - (learning_rate * dw)
        b = b - (learning_rate * db)
        
        if i % (iterations/5) == 0:
            print(f"Iteration {i}: Cost {cost:.4f}, Weight {w:.4f}, Bias {b:.4f}")
            
    return w, b

### Executing the function with different parameters <a id = "dp">

#### Increasing iterations while keeping at learning rate 0.01

In [32]:
#10 Iterations, 0.01 Learning Rate
w_final, b_final = gradient_descent(X, Y, 10, 0.01)
print(f"\nFinal Result: y = {w_final:.2f}x + {b_final:.2f}")

Iteration 0: Cost 89.0000, Weight 0.6200, Bias 0.1800
Iteration 2: Cost 30.8319, Weight 1.4532, Bias 0.4272
Iteration 4: Cost 11.0700, Weight 1.9371, Bias 0.5776
Iteration 6: Cost 4.3508, Weight 2.2173, Bias 0.6715
Iteration 8: Cost 2.0610, Weight 2.3789, Bias 0.7323

Final Result: y = 2.43x + 0.75


In [34]:
#100 Iterations, 0.01 Learning Rate
w_final, b_final = gradient_descent(X, Y, 100, 0.01)
print(f"\nFinal Result: y = {w_final:.2f}x + {b_final:.2f}")

Iteration 0: Cost 89.0000, Weight 0.6200, Bias 0.1800
Iteration 20: Cost 0.8221, Weight 2.5770, Bias 0.8817
Iteration 40: Cost 0.7164, Weight 2.5476, Bias 1.0228
Iteration 60: Cost 0.6257, Weight 2.5118, Bias 1.1523
Iteration 80: Cost 0.5464, Weight 2.4783, Bias 1.2733

Final Result: y = 2.45x + 1.38


In [35]:
#1000 Iterations, 0.01 Learning Rate
w_final, b_final = gradient_descent(X, Y, 1000, 0.01)
print(f"\nFinal Result: y = {w_final:.2f}x + {b_final:.2f}")

Iteration 0: Cost 89.0000, Weight 0.6200, Bias 0.1800
Iteration 200: Cost 0.2424, Weight 2.3186, Bias 1.8499
Iteration 400: Cost 0.0625, Weight 2.1618, Bias 2.4158
Iteration 600: Cost 0.0161, Weight 2.0822, Bias 2.7032
Iteration 800: Cost 0.0042, Weight 2.0418, Bias 2.8493

Final Result: y = 2.02x + 2.92


#### Increasing Learning Rate will keeping iterations at 100

In [44]:
#100 Iterations, 0.03 Learning Rate
w_final, b_final = gradient_descent(X, Y, 100, 0.03)
print(f"\nFinal Result: y = {w_final:.2f}x + {b_final:.2f}")

Iteration 0: Cost 89.0000, Weight 1.8600, Bias 0.5400
Iteration 20: Cost 0.6248, Weight 2.5080, Bias 1.1661
Iteration 40: Cost 0.4156, Weight 2.4143, Bias 1.5043
Iteration 60: Cost 0.2764, Weight 2.3379, Bias 1.7802
Iteration 80: Cost 0.1838, Weight 2.2755, Bias 2.0052

Final Result: y = 2.23x + 2.18


In [45]:
#100 Iterations, 0.06 Learning Rate
w_final, b_final = gradient_descent(X, Y, 100, 0.06)
print(f"\nFinal Result: y = {w_final:.2f}x + {b_final:.2f}")

Iteration 0: Cost 89.0000, Weight 3.7200, Bias 1.0800
Iteration 20: Cost 0.4138, Weight 2.4092, Bias 1.5228
Iteration 40: Cost 0.1823, Weight 2.2716, Bias 2.0195
Iteration 60: Cost 0.0803, Weight 2.1803, Bias 2.3492
Iteration 80: Cost 0.0354, Weight 2.1196, Bias 2.5681

Final Result: y = 2.08x + 2.71


In [46]:
#100 Iterations, 0.09 Learning Rate
w_final, b_final = gradient_descent(X, Y, 100, 0.09)
print(f"\nFinal Result: y = {w_final:.2f}x + {b_final:.2f}")

Iteration 0: Cost 89.0000, Weight 5.5800, Bias 1.6200
Iteration 20: Cost 11516.9833, Weight 36.2927, Bias 11.2202
Iteration 40: Cost 1506173.7515, Weight 390.5868, Bias 109.9433
Iteration 60: Cost 196979784.3323, Weight 4443.9389, Bias 1232.9745
Iteration 80: Cost 25761328950.9274, Weight 50798.8731, Bias 14072.7191

Final Result: y = -514273.88x + -142443.28


Shows that too high a learning rate will cause the numbers to "ping pong" back and forth at increasing step sizes, getting further away from the target.

### Best Result: (In this Scenario)
Iterations = 1000, Learning Rate = 0.01

Final Result: y = 2.02x + 2.92